crates/noirc_evaluator/src/ssa/acir_gen/internal_var_cache.rs

```diff
@@ -31,8 +37,24 @@ impl InternalVarCache {
 
         let mut var = match ctx.try_get_node(id)? {
             NodeObject::Const(c) => {
+                // use the InternalVar from constants if exists
                 let field_value = FieldElement::from_be_bytes_reduce(&c.value.to_bytes_be());
-                InternalVar::from_constant(field_value)
+                let mut result = InternalVar::from_constant(field_value);
+                if let Some(c_var) = self.constants.get(&field_value) {
+                    result = c_var.clone();
+                }
+                //use witness from other nodes if exists
+                if let Some(constant_ids) = ctx.constants.get(&field_value) {
+                    for c_id in constant_ids {
+                        if let Some(i_var) = self.inner.get(c_id) {
+                            if i_var.cached_witness().is_some() {
```

Comment:
In what cases would we call `set_witness` with `None` ?

Reply:
If it is not called, witness will be None.

Reply:
As discussed, I have changed the set_witness to take witness as input instead of an Option.